pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,527 @@
+
+	nodeInformer     coreinformers.NodeInformer
+	nodeLister       corelisters.NodeLister
+	nodeListerSynced cache.InformerSynced
+
+	ipPoolInformer          crdinformers.ExternalIPPoolInformer
+	ipPoolLister            crdlister.ExternalIPPoolLister
+	ipPoolInformerHasSynced cache.InformerSynced
+}
+
+// NewCluster return a new *Cluster
+func NewCluster(
+	clusterBindPort int,
+	nodeInformer coreinformers.NodeInformer,
+	nodeConfig *config.NodeConfig,
+	ipPoolInformer crdinformers.ExternalIPPoolInformer,
+) (*Cluster, error) {
+	// The Node join/leave events will be notified via it.
+	nodeEventCh := make(chan memberlist.NodeEvent, 1024)
+	c := &Cluster{
+		bindPort:                clusterBindPort,
+		nodeInformer:            nodeInformer,
+		nodeLister:              nodeInformer.Lister(),
+		nodeListerSynced:        nodeInformer.Informer().HasSynced,
+		NodeConfig:              nodeConfig,
+		nodeName:                nodeConfig.Name,
+		nodeEventsCh:            nodeEventCh,
+		ipPoolInformer:          ipPoolInformer,
+		ipPoolLister:            ipPoolInformer.Lister(),
+		ipPoolInformerHasSynced: ipPoolInformer.Informer().HasSynced,
+		nodeSelectors:           make(map[string]*metav1.LabelSelector),
+		localNodeEventCh:        make(chan localNodeEventType, 1),
+	}
+	c.conHash = newNodeConsistentHashMap()
+
+	bindPort := c.bindPort
+	hostIP := c.NodeConfig.NodeIPAddr.IP
+
+	nodeMember := fmt.Sprintf("%s:%d", hostIP.String(), bindPort)
+
+	klog.V(2).Infof("Add new node: %s", nodeMember)
+
+	conf := memberlist.DefaultLocalConfig()
+	conf.Name = c.nodeName
+
+	conf.BindPort = bindPort
+	conf.AdvertisePort = bindPort
+	conf.Events = &memberlist.ChannelEventDelegate{Ch: nodeEventCh}
+
+	klog.V(1).Infof("Memberlist cluster configs: %+v", conf)
+
+	mList, err := memberlist.Create(conf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create memberlist: %v", err.Error())
+	}
+
+	c.mList = mList
+	c.existingMembers = sets.NewString(nodeMember)
+
+	nodeInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addNodeHandler,
+			UpdateFunc: c.updateNodeHandler,
+			DeleteFunc: c.deleteNodeHandler,
+		},
+		resyncPeriod,
+	)
+
+	ipPoolInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.addIPPoolHandler,
+			UpdateFunc: c.updateIPPoolHandler,
+			DeleteFunc: c.deleteIPPoolHandler,
+		},
+		resyncPeriod,
+	)
+	return c, nil
+}
+
+func (c *Cluster) addIPPoolHandler(obj interface{}) {
+	ipPool, ok := obj.(*v1alpha2.ExternalIPPool)
+	if !ok {
+		klog.Errorf("Add ExternalIPPool callback error, unexpected object type: %v", obj)
+		return
+	}
+	c.addOrUpdateNodeSelector(ipPool)
+	shouldJoin := c.shouldJoinCluster(c.nodeName)
+	joined := c.localNodeJoined()
+	klog.V(0).Infof("Processing ExternalIPPool %s create event, joined (%t), should join: %t", ipPool.Name, joined, shouldJoin)
+	if !joined && shouldJoin {
+		c.localNodeEventCh <- localNodeWillJoin
```

Comment:
Maybe the common controller pattern can make this simpler:
Normally we can defer more work to the goroutine that consumes `localNodeEventCh`. The benefits are:
1.  Reduce code redundancy in the 6 eventHandlers, for example `c.shouldJoinCluster` and `c.localNodeJoined()`.
2.  Have a single place to calculate what to do based on latest externalIPPool and Node cache.
3. Merging several events to avoid duplicate processing. For example, if several IPPool events and Node events come at the same time, the goroutine only needs to process once.

It usually uses a workqueue but in your case, maybe a channel with 1 buffer is enough.

Reply:
done